### PR TITLE
Preserve original GraphArgs for shape guard codegen

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #90665
* #90664

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

### CODE REVIEW

File: test/dynamo/test_repros.py [hunk 1]
```diff
@@ -1959,6 +1959,77 @@ def f(x, y):
...
+        # https://github.com/pytorch/torchdynamo/issues/1925 ).  So long
+        # as Dynamo eagerly guards in this way, we have an invariant that
+        # all locals are guaranteed to show up in GraphArgs before the
+        # inlining function call, in which case we will always have enough
+        # information to codegen our guards so long as we don't prune the
+        # unused GraphArgs away (and indeed, the direct fix for this bug
+        # was to make sure we use original GraphArgs).  Non locals,
+        # conversely, typically are static, and so won't have guards allocated
+        # for them.  That being said, there may still be a way to trigger
+        # this error.
```
Comment (voznesenskym) [inline test/dynamo/test_repros.py:2016]: This deserves its own test to protect this invariant, we take it as a matter of course for now, but it changing subtly under the hood could cause really nasty to debug issues. Let's file an issue to test this (specifically, the exhaustive guarding of locals).

File: torch/_dynamo/output_graph.py [hunk 2]
```diff
@@ -180,6 +180,14 @@ def __init__(
...
         self.graph = torch.fx.Graph()
         self.graphargs: List[GraphArg] = []
+        # Although we prune unused graphargs before sending graphs to
+        # compilers, we may have legitimately triggered shape guards
+        # on "unused" inputs that we must keep track of.  So after
+        # remove_unused_graphargs is called, orig_graphargs and
+        # graphargs no longer alias; orig_graphargs is the original
+        # graphargs, and graphargs is the pruned list.  Guard creation
+        # should use original graphargs.
+        self.orig_graphargs: List[GraphArg] = self.graphargs
```
Comment (voznesenskym) [inline torch/_dynamo/output_graph.py:190]: Are there memory implications for keeping around 2x of everything?

**Review by voznesenskym — APPROVED** [review]

File: test/dynamo/test_repros.py (hunk 1 above)
Comment (ezyang) [reply test/dynamo/test_repros.py:2016]: This isn't a variant. I just failed to come up with a test case that could actually trigger the more naughty fix.

The proper fix is to checkpoint shape env properly. But I'm hoping we can put that off a bit.

Comment (ezyang) [reply test/dynamo/test_repros.py:2016]: It turns out that you can trigger the edge case discussed here with unspecialized ints, and in fact quite easily.

**Comment by bhack** [comment]
I had an error on nightly mentioning this PR
```python
torch._dynamo.exc.BackendCompilerFailed: backend='inductor' raised:
AssertionError: s6 (could be from ["L['self'].last_size_2d[0]"]) not in {s1: ["L['t0']"], s10: ["L['t1']"], s7: ["L['t3']"], s6: []}.  If this assert is failing, it could be due to the issue described in https://github.com/pytorch/pytorch/pull/90665

Set TORCH_LOGS="+dynamo" and TORCHDYNAMO_VERBOSE=1 for more information
```
What info do you need?

**Comment by ezyang** [comment]
Are you running the cache by any chance? If so, it's probably https://github.com/pytorch/pytorch/issues/127970

**Comment by bhack** [comment]
Yes it is with `TORCHINDUCTOR_FX_GRAPH_CACHE=1`

**Comment by brthor** [comment]
I'm commenting here for anyone else who lands here by the link in the exception.

I hit this in pytorch `3.3.0`.

The issue was that I was slicing the input tensor to my network that was torch.compiled, which caused the stride to mismatch the shape.

I was doing this:
```
x = x[:, :, :new_size]
net(x)
```

Adding .contiguous() fixed this issue:
```
x = x[:, :, :new_size].contiguous()
net(x)
```


```
E0630 04:11:58.231000 139761003709504 torch/_guards.py:251] [1/1] Error while creating guard:
E0630 04:11:58.231000 139761003709504 torch/_guards.py:251] [1/1] Name: ''
E0630 04:11:58.231000 139761003709504 torch/_guards.py:251] [1/1]     Source: shape_env
E0630 04:11:58.231000 139761003709504 torch/_guards.py:251] [1/1]     Create Function: SHAPE_ENV
E0630 04:11:58.231000 139761003709504 torch/_guards.py:251] [1/1]     Guard Types: None
E0630 04:11:58.231000 139761003709504 torch/_guards.py:251] [1/1]     Code List: None
E0630 04:11:58.231000 139761003709504 torch/_guards.py:251] [1/1]     Object Weakref: None
E0630 04:11:58.231000 139761003709504 torch/_guards.py:251] [1/1]     Guarded Class Weakref: None
E0630 04:11:58.231000 139761003709504 torch/_guards.py:253] [1/1] Created at:
E0630 04:11:58.231000 139761003709504 torch/_guards.py:253] [1/1]   File "/opt/conda/lib/python3.10/site-packages/torch/_dynamo/convert_frame.py", line 482, in transform
E0630 04:11:58.231000 139761003709504 torch/_guards.py:253] [1/1]     tracer = InstructionTranslator(
E0630 04:11:58.231000 139761003709504 torch/_guards.py:253] [1/1]   File "/opt/conda/lib/python3.10/site-packages/torch/_dynamo/symbolic_convert.py", line 2060, in __init__
E0630 04:11:58.231000 139761003709504 torch/_guards.py:253] [1/1]     output=OutputGraph(
E0630 04:11:58.231000 139761003709504 torch/_guards.py:253] [1/1]   File "/opt/conda/lib/python3.10/site-packages/torch/_dynamo/output_graph.py", line 310, in __init__
E0630 04:11:58.231000 139761003709504 torch/_guards.py:253] [1/1]     self.init_ambient_guards()
E0630 04:11:58.231000 139761003709504 torch/_guards.py:253] [1/1]   File "/opt/conda/lib/python3.10/site-packages/torch/_dynamo/output_graph.py", line 421, in init_ambient_guards
E0630 04:11:58.231000 139761003709504 torch/_guards.py:253] [1/1]     self.guards.add(ShapeEnvSource().make_guard(GuardBuilder.SHAPE_ENV))

...

AssertionError: s2 (could be from ["L['var_x']._base.stride()[0]"]) not in {s1: [], s2: []}.  If this assert is failing, it could be due to the issue described in https://github.com/pytorch/pytorch/pull/90665
```

I think there is probably a bug here though.

**Comment by ezyang** [comment]
Oh, for yours, it's possible nightly has fixed it. I think I need to update the error description here
